File: api/natal.py
from http.server import BaseHTTPRequestHandler
import json, sys, os
from datetime import datetime
from urllib.parse import urlparse, parse_qs

sys.path.insert(0, os.path.dirname(__file__))
from _astro import calc_positions, geocode, sb_get_user

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self._cors()
        self.end_headers()

    def do_GET(self):
        try:
            params = parse_qs(urlparse(self.path).query)
            telegram_id = int(params.get('id', [0])[0])
            if not telegram_id:
                return self._json({"error": "id required"}, 400)

            user = sb_get_user(telegram_id)
            if not user:
                return self._json({"error": "Пользователь не найден"}, 404)

            bdate = datetime.strptime(user['birthdate'], "%Y-%m-%d")
            btime = datetime.strptime(user.get('birthtime', '12:00:00'), "%H:%M:%S")

            lat = user.get('birth_lat') or 0
            lon_c = user.get('birth_lon') or 0
            if (not lat or lat == 55.7558) and user.get('birthplace'):
                lat, lon_c = geocode(user['birthplace'])

            positions = calc_positions(bdate.year, bdate.month, bdate.day, btime.hour, btime.minute, lat, lon_c)
            self._json({
                "name": user['name'],
                "birthdate": user['birthdate'],
                "birthplace": user.get('birthplace'),
                "planets": [{"name": v['name_ru'], "sign": v['sign'], "degree": v['degree']} for v in positions.values()]
            })
        except Exception as e:
            self._json({"error": str(e)}, 500)

    def _cors(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')

    def _json(self, data, status=200):
        body = json.dumps(data, ensure_ascii=False).encode('utf-8')
        self.send_response(status)
        self.send_header('Content-Type', 'application/json; charset=utf-8')
        self._cors()
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, *args):
        pass

File: api/natal.js
const { getUser, calcPlanets, signFromLong, sbFetch } = require('./_shared')

module.exports = async (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*')
  if (req.method === 'OPTIONS') return res.status(200).end()

  try {
    const telegramId = parseInt(req.query.id)
    if (!telegramId) return res.status(400).json({ error: 'id required' })

    const user = await getUser(telegramId)
    if (!user) return res.status(404).json({ error: 'User not found' })

    // Проверяем кэш в Supabase
    const cached = await sbFetch(`/natal_charts?user_id=eq.${user.id}&limit=1`)
    if (cached?.[0]) return res.json(cached[0])

    // Рассчитываем натальную карту
    const [y, m, d] = user.birth_date.split('-').map(Number)
    const [h, min] = (user.birth_time || '12:00').split(':').map(Number)
    const birthDate = new Date(Date.UTC(y, m - 1, d, h, min))

    const planets = calcPlanets(birthDate)
    const sun_sign = planets.Sun?.sign || 'Овен'
    const moon_sign = planets.Moon?.sign || 'Овен'
    // Асцендент без точного места рождения = примерный (через время)
    const ascLon = (birthDate.getUTCHours() / 24 * 360 + planets.Sun?.longitude || 0) % 360
    const ascendant = signFromLong(ascLon)

    const chart = {
      user_id: user.id,
      birth_date: user.birth_date,
      birth_time: user.birth_time,
      birth_place: user.birth_place,
      sun_sign,
      moon_sign,
      ascendant,
      planets_json: planets,
    }

    // Сохраняем в Supabase
    const rows = await sbFetch('/natal_charts', {
      method: 'POST',
      prefer: 'return=representation',
      body: JSON.stringify(chart),
    })

    res.json(rows?.[0] || chart)
  } catch (e) {
    res.status(500).json({ error: e.message })
  }
}
